### DISCUSSION TITLE
Fetching data periodically using nextjs server actions

### DISCUSSION THREAD

**Discussion by myhendry** [vercel/next.js#72012] [Help]
Hi, I'm currently using server actions to fetch data from my prisma mongo database in server component. For authentication, I am using kinde 

I like to do a periodic refresh to fetch the latest data to update the UI. I'm aware I can use fetch with revalidation on the client side but for my app's use case, it is better to use server components

I can't use fetch in  my server component to call my own api routes as I understand it is not good practise and also, if I call my own api route using fetch from my server component, kinde auth is not protecting the route

Any expert has any idea how I can solve it? Thanks

**Comment by robahtou** [comment]
Using RSA does not automatically cache your data. You have to construct your server action to have it cache the results. SInce it appears you are not doing this everytime you call your server action it will make a fresh database call and get the latest data.

**Reply by myhendry** [reply]
@robahtou Using RSA works fine when the server component page rendered the first time. However, what if I like to get the latest data using RSA every 5 seconds automatically? Similar to using fetch and revalidation that calls the api every x seconds but using server actions? Is that even possible?

**Reply by robahtou** [reply]
Using `[revalidate](https://nextjs.org/docs/canary/app/building-your-application/caching#time-based-revalidation)` means that the data is cached and you want to purge the data after a certain amount of time.

If you have a `fetch` call that is getting data automatically after a x seconds, then you are probably wrapping that `fetch` call in a function that is constructed to call your `fetch` after x seconds.

```javascript
// something like this
async function temporalFetch(seconds) {
  const fetchData = async () => {
    // not cached -- every call returns fresh data (assuming latest updates from nextjs)
    const response = await fetch('/api/my_api/get_something');
    const data = await response.json();
   setData(data); // assuming you are using `useState`
  };

  setInterval(fetchData, seconds);
}
```
In the similar vain, for RSA, in a client component you would wrap the RSA in a function (`useEffect`?) that runs the RSA after x seconds. Calling a RSA does not automatically refresh a page nor does it automatically revalidate the data that another page might be using from the RSA call. Here are some resources on revalidating:

* https://nextjs.org/docs/canary/app/api-reference/functions/revalidatePath
* https://nextjs.org/docs/canary/app/api-reference/functions/revalidateTag
* https://nextjs.org/docs/canary/app/api-reference/directives/use-cache (new!)

Essentially, the client needs to `pull` data from the server. Therefore, you need to construct a function on the client that `pulls` data every x seconds, and properly update the UI with react hooks that are reactive to your data fetching.

**Reply by myhendry** [reply]
@robahtou using fetch + revalidate on the server and using fetch + revalidate on the client, I am aware. It's just that I'm directly calling my mongo database using prisma inside my server component and I was wondering if I could do something like fetch + revalidate where I can call my database at periodic intervals and update my UI? Sorry, I was not clear just now

**Reply by robahtou** [reply]
I'm having trouble understanding what you're trying to do. Let's use this example to get on the same page.

```jsx
// server component
async function NoteList({searchText}) {
  const notes = await mongoPrisma.someMethodToGetData();

  return (
    <ul className="notes-list">
      {notes.map((note) => (
        <li key={note.id}>
          {note.title}
        </li>
      ))}
    </ul>
  );
}
```

This server component is calling directly to mongo. My question to you, how are you calling this component?

**Reply by myhendry** [reply]
Hi, so sorry for the confusion. I'm able to get the posts. 

I was wondering if I can do periodic calls to the server to get the latest data to update the UI just like using fetch + revalidation but directly calling the database through prisma mongo instead?

// Inside Server component

const posts = await prisma.post.findMany({
   where: {
     private: false
   }
});

**Comment by icyJoseph** [comment]
Check this Deep Dive, https://github.com/vercel/next.js/discussions/54075:

It presents these component (copy-pasting for visibility):

```tsx
import { useRouter, usePathname } from 'next/navigation'
import { debounce } from 'lodash';

interface DataRevalidatorProps {
  /**
   * The timeout in milliseconds to wait before revalidating the data.
   */
  timeout?: number;
}

export function DataRevalidator({
  timeout = 5000,
}: DataRevalidatorProps): null {
  const router = useRouter();
  const pathname = usePathname()

  const getRefresher = useCallback(() => debounce(
    () => {
      // Can be removed, just to showcase when it triggers refresh
      console.log(
        `[${new Date().toLocaleTimeString()}] Triggering router.refresh()...`,
      );
      router.refresh();
    },
    timeout,
		{ leading: false, trailing: true }
  ), [router, timeout]);

  // Trigger router.refresh() whenever `pathname` changes.
	// Note: this does not cover searchParams, you can add `useSearchParams` if needed.
  useEffect(()=> {
		const refresher = getRefresher()
    refresher()
	}, [revalidate, pathname])

  // Trigger router.refresh() when focus goes back to the window (similar to useSWR / react-query)
  useEffect(() => {
    if(process.env.NODE_ENV === 'development') {
			return
    }

		const refresher = getRefresher()

    window.addEventListener('focus', refresher);
    return () => {
      refresher.cancel();
      window.removeEventListener('focus', refresher);
    };
  }, [revalidate]);

  return null;
}
```

Read through that post, and see if that DataRevalidator is good enough for you.

**Comment by aniruddhaadak80** [comment]
Hi @myhendry ,

To fetch data periodically in a Next.js server component while ensuring proper authentication and leveraging server actions, consider the following approaches:

1. **Use Server Actions with Revalidation**:
   - Implement server actions that fetch data from your Prisma Mongo database. Use Next.js’s built-in revalidation options to periodically refresh the data without needing to rely on client-side fetching.
   ```javascript
   export const fetchData = async () => {
     const data = await prisma.collection.findMany();
     return data;
   };
   ```

2. **Set Up a Polling Mechanism**:
   - Use `setInterval` in a `useEffect` hook within a client component to periodically call the server action, ensuring it fetches the latest data.
   ```javascript
   import { useEffect, useState } from 'react';

   const YourComponent = () => {
     const [data, setData] = useState([]);

     useEffect(() => {
       const interval = setInterval(async () => {
         const response = await fetch('/api/your-endpoint', {
           method: 'GET',
           headers: {
             'Authorization': `Bearer ${yourAuthToken}`, // Include auth token if needed
           },
         });
         const newData = await response.json();
         setData(newData);
       }, 5000); // Fetch data every 5 seconds

       return () => clearInterval(interval); // Cleanup on unmount
     }, []);

     return (
       <div>{JSON.stringify(data)}</div>
     );
   };
   ```

3. **Server-Side Caching with Middleware**:
   - If you need to ensure Kinde auth protects your data, consider using middleware to handle authentication and caching server-side. This allows you to cache the results of your server action and refresh them at defined intervals.
   ```javascript
   export const middleware = async (req, res, next) => {
     // Check Kinde authentication
     const isAuthenticated = await checkKindeAuth(req);
     if (!isAuthenticated) return res.status(401).end();

     // Fetch and cache data here
     const cachedData = await fetchCachedData();

     // Set a revalidation strategy, e.g., cache for 1 minute
     res.setHeader('Cache-Control', 's-maxage=60, stale-while-revalidate=30');
     res.json(cachedData);
   };
   ```

Implementing these strategies will allow you to periodically refresh data from your Mongo database while maintaining proper authentication and utilizing Next.js server components effectively.


Thanks .

**Reply by myhendry** [reply]
I see. Thanks. Looks like I can run it on client side only

**Reply by robahtou** [reply]
This is obviously AI generated response slop.

`Use Next.js’s built-in revalidation options to periodically refresh the data without needing to rely on client-side fetching.` And shows an example that does not do that. `polling mechanism` is not server action implemented. And then using middleware, auth, cachedData. 

If this answers your question I was way off mark of what you were trying to do.

**Reply by myhendry** [reply]
@robahtou oh, thanks for the spot. I watched some YouTube videos and is aware how to do RSA and revalidatePath on a form submission. 

I am also clear can use nextjs fetch + revalidate to refetch data from api on both client and server side. 

However, I read somewhere it's not advisable to use fetch + revalidate to get data from own api. Also, as I'm using kinde auth, I cannot run fetch + revalidate from a server component to my own nextjs api as it's using JWT so kinde auth doesn't work when I call fetch from my server component. 

Thus, I was wondering if it's technically possible to clear the cache and give my users an updated list of data whenever they access the page again? 

Sorry, I'm confused by nextjs 14 caching. I'm still trying to get my head around it. I understand Nextjs 15 will not have default cache so it might resolve my issue
